File: tests/test_budget.py
"""Tests for the budget-limit logic: can_afford() and place_order()."""

import pytest

import db


@pytest.fixture
def shop_db(tmp_path, monkeypatch):
    monkeypatch.setattr(db, "SHOP_DB_PATH", tmp_path / "test_shop.sqlite")
    db.init_shop_db()
    return db


def test_can_afford_allows_order_that_exactly_uses_remaining_budget():
    assert db.can_afford(budget=100, spent=60, order_total=40)


def test_can_afford_blocks_order_that_exceeds_remaining_budget():
    assert not db.can_afford(budget=100, spent=60, order_total=40.01)


def test_place_order_records_an_order_and_updates_spend(shop_db):
    product = shop_db.get_products()[0]

    result = shop_db.place_order("test_user", product["id"], budget=1_000_000)

    assert result["success"]
    assert shop_db.get_spent("test_user") == product["price"]
    assert len(shop_db.get_orders("test_user")) == 1


def test_place_order_refuses_an_order_over_budget_and_does_not_record_it(shop_db):
    product = shop_db.get_products()[0]

    result = shop_db.place_order("test_user", product["id"], budget=product["price"] - 1)

    assert not result["success"]
    assert shop_db.get_spent("test_user") == 0
    assert len(shop_db.get_orders("test_user")) == 0


def test_place_order_allows_an_order_exactly_at_the_budget_boundary(shop_db):
    product = shop_db.get_products()[0]

    result = shop_db.place_order("test_user", product["id"], budget=product["price"])

    assert result["success"]


def test_place_order_rejects_an_unknown_product(shop_db):
    result = shop_db.place_order("test_user", product_id=999999, budget=1_000_000)

    assert not result["success"]

File: tests/testthat/test-budget.R
test_that("can_afford allows an order that exactly uses the remaining budget", {
  expect_true(can_afford(budget = 100, spent = 60, order_total = 40))
})

test_that("can_afford blocks an order that exceeds the remaining budget", {
  expect_false(can_afford(budget = 100, spent = 60, order_total = 40.01))
})

test_that("place_order records an order and updates spend", {
  con <- get_shop_con(":memory:")
  on.exit(dbDisconnect(con))
  init_shop_db(con)

  product <- get_products(con)[1, ]
  result <- place_order(con, "test_user", product$id, quantity = 2, budget = 1e6)

  expect_true(result$success)
  expect_equal(get_spent(con, "test_user"), product$price * 2)
  expect_equal(nrow(get_orders(con, "test_user")), 1)
})

test_that("place_order refuses an order over budget and does not record it", {
  con <- get_shop_con(":memory:")
  on.exit(dbDisconnect(con))
  init_shop_db(con)

  product <- get_products(con)[1, ]
  result <- place_order(
    con, "test_user", product$id, quantity = 1, budget = product$price - 1
  )

  expect_false(result$success)
  expect_equal(get_spent(con, "test_user"), 0)
  expect_equal(nrow(get_orders(con, "test_user")), 0)
})

test_that("place_order rejects an unknown product", {
  con <- get_shop_con(":memory:")
  on.exit(dbDisconnect(con))
  init_shop_db(con)

  result <- place_order(con, "test_user", product_id = 9999, quantity = 1, budget = 1e6)

  expect_false(result$success)
})

test_that("place_order rejects a quantity below 1", {
  con <- get_shop_con(":memory:")
  on.exit(dbDisconnect(con))
  init_shop_db(con)

  product <- get_products(con)[1, ]
  result <- place_order(con, "test_user", product$id, quantity = 0, budget = 1e6)

  expect_false(result$success)
})
